src/server/stream_family.cc

```diff
@@ -1192,97 +1240,245 @@ void StreamFamily::XRevRange(CmdArgList args, ConnectionContext* cntx) {
           kSyntaxErr);
+      return std::nullopt;
     }
 
-    ParsedStreamId id;
     if (!ParseID(idstr, true, 0, &id)) {
-      return (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+      (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+      return std::nullopt;
     }
+
     // We only include messages with IDs greater than start so increment the
     // starting ID.
     streamIncrID(&id.val);
-    read_opts.stream_ids.emplace(key, id);
+    opts.stream_ids.emplace(key, id);
   }
 
-  unsigned shard_count = shard_set->size();
-  vector<vector<RecordVec>> xread_resp(shard_count);
+  return opts;
+}
+
+// Returns the last ID of each stream in the transaction.
+OpResult<unordered_map<string_view, streamID>> StreamLastIDs(Transaction* trans) {
+  vector<OpResult<vector<pair<string_view, streamID>>>> last_ids_res(shard_set->size());
+
   auto cb = [&](Transaction* t, EngineShard* shard) {
     ShardId sid = shard->shard_id();
-    xread_resp[sid] = OpRead(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()), read_opts);
+    last_ids_res[sid] = OpLastIDs(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()));
     return OpStatus::OK;
   };
-  OpStatus result = cntx->transaction->ScheduleSingleHop(std::move(cb));
-  CHECK_EQ(OpStatus::OK, result);
+  trans->Execute(std::move(cb), false);
+
+  unordered_map<string_view, streamID> last_ids;
+  for (auto res : last_ids_res) {
+    if (!res) {
+      return res.status();
+    }
+
+    for (auto& e : *res) {
+      last_ids.emplace(e.first, e.second);
+    }
+  }
+  return last_ids;
+}
+
+void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
+  auto opts = ParseReadArgsOrReply(args, cntx);
+  if (!opts) {
+    return;
+  }
+
+  cntx->transaction->Schedule();
+
+  auto last_ids = StreamLastIDs(cntx->transaction);
+  if (!last_ids) {
+    // Close the transaction.
+    auto close_cb = [&](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
+    cntx->transaction->Execute(std::move(close_cb), true);
+
+    if (last_ids.status() == OpStatus::WRONG_TYPE) {
+      (*cntx)->SendError(kWrongTypeErr);
+      return;
+    }
+
+    return (*cntx)->SendNullArray();
+  }
+
+  // Resolve '$' IDs and check if there are any streams with entries that can
+  // be resolved without blocking.
+  bool block = true;
+  for (auto& [stream, requested_id] : opts->stream_ids) {
+    if (auto last_id_it = last_ids->find(stream); last_id_it != last_ids->end()) {
+      streamID last_id = last_id_it->second;
+
+      // Resolve $ to the last ID in the stream.
+      if (requested_id.last_id) {
+        requested_id.val = last_id;
+        // We only include messages with IDs greater than the last message so
+        // increment the ID.
+        streamIncrID(&requested_id.val);
+        requested_id.last_id = false;
+        continue;
+      }
+
+      if (streamCompareID(&last_id, &requested_id.val) >= 0) {
+        block = false;
+      }
+    }
+  }
+
+  if (block) {
```

Comment:
is it possible to factor out the code inside `if (block) {` into a separate function?
the flow in this function is too long.

Reply:
yep updated